File: LeetCode/1762.Buildings_With_an_Ocean_View/1762.Buildings_With_an_Ocean_View.py
#!/usr/bin/env python3


class Solution:
    def findBuildings(self, heights: List[int]) -> List[int]:
        # height[i] > suffix[i+1]
        n = len(heights)
        suffix = [0] * (n + 1)
        for i in range(n - 1, -1, -1):
            suffix[i] = max(suffix[i + 1], heights[i])

        res = []
        for i, h in enumerate(heights):
            if h > suffix[i + 1]:
                res.append(i)

        return res


class Solution:
    def findBuildings(self, heights: List[int]) -> List[int]:
        # height[i] > suffix[i+1]
        n = len(heights)
        res = []
        max_height = 0
        for i in range(n - 1, -1, -1):
            if heights[i] > max_height:
                res.append(i)
                max_height = heights[i]
        return res[::-1]

File: LeetCode/1762.Buildings_With_an_Ocean_View/1762.Buildings_With_an_Ocean_View.swift
class Solution {
    func findBuildings(_ heights: [Int]) -> [Int] {
        // scan backforward, store highest height, compare it
        // if lower than it, so no ocean view, if higher than it, has ocean view
        let n = heights.count
        var res: [Int] = []

        var height = 0
        for i in stride(from: n - 1, to: -1, by: -1) {
            if heights[i] > height {
                res.append(i)
                height = heights[i]
            }
        }

        res.reverse()

        return res
    }
}

// monotonic stack solution, support dynamic add building
class Solution {
    func findBuildings(_ heights: [Int]) -> [Int] {
        // monotonic stack: decreasing order
        // scan forward, if heights[i] > last element in stack, stack pop

        let n = heights.count
        var stack: [Int] = []

        for i in 0 ..< n {
            while let last = stack.last, heights[i] >= heights[last] {
                stack.removeLast()
            }
            stack.append(i)
        }

        return stack
    }
}
